File: android/src/main/kotlin/io/vezz/flutter_local_ai/FlutterLocalAiPlugin.kt
package io.vezz.flutter_local_ai

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel

/**
 * Registration for flutter_local_ai on Android.
 *
 * All behaviour lives in [LocalAiSessionService]: one pigeon host API and one
 * event channel, driven from Dart by both the prompt-oriented `FlutterLocalAi`
 * facade and the session API. There is deliberately no second method channel —
 * a parallel implementation of the same generation logic is where the two
 * surfaces drift apart.
 */
class FlutterLocalAiPlugin : FlutterPlugin {
  private var sessionService: LocalAiSessionService? = null
  private var events: EventChannel? = null

  override fun onAttachedToEngine(
    @NonNull binding: FlutterPlugin.FlutterPluginBinding
  ) {
    val service = LocalAiSessionService(binding.applicationContext)
    sessionService = service
    LocalAiService.setUp(binding.binaryMessenger, service)
    events = EventChannel(binding.binaryMessenger, "flutter_local_ai_events")
      .apply { setStreamHandler(service) }
  }

  override fun onDetachedFromEngine(
    @NonNull binding: FlutterPlugin.FlutterPluginBinding
  ) {
    LocalAiService.setUp(binding.binaryMessenger, null)
    events?.setStreamHandler(null)
    events = null
    sessionService?.cleanup()
    sessionService = null
  }
}

File: darwin/flutter_local_ai/Classes/FlutterLocalAiPlugin.swift
import Foundation

#if os(OSX)
  @preconcurrency import FlutterMacOS
#elseif os(iOS)
  @preconcurrency import Flutter
#endif

/// Registration for flutter_local_ai on iOS and macOS.
///
/// All behaviour lives in `LocalAiSessionService`: one pigeon host API and one
/// event channel, driven from Dart by both the prompt-oriented
/// `FlutterLocalAi` facade and the session API. There is deliberately no
/// second method channel — a parallel implementation of the same generation
/// logic is where the two surfaces drift apart.
@objc public class FlutterLocalAiPlugin: NSObject, FlutterPlugin {
  /// Retains the service for the lifetime of the process. Pigeon's `setUp`
  /// keeps only a weak reference to the handler, and `FlutterEventChannel`
  /// does not retain its stream handler either, so without this the service
  /// would deallocate as soon as `register` returned and every call would
  /// fail with a channel error.
  private static var sessionService: LocalAiSessionService?

  public static func register(with registrar: FlutterPluginRegistrar) {
    #if os(OSX)
      let messenger = registrar.messenger
    #elseif os(iOS)
      let messenger = registrar.messenger()
    #endif

    let service = LocalAiSessionService(
      toolRunner: LocalAiToolRunner(binaryMessenger: messenger))
    sessionService = service
    LocalAiServiceSetup.setUp(binaryMessenger: messenger, api: service)
    FlutterEventChannel(name: "flutter_local_ai_events", binaryMessenger: messenger)
      .setStreamHandler(service)
  }
}
